### PR TITLE
BUG: Fix datetime->string conversion

### LINKED ISSUE #9712: Casting datetime to string truncates modulo 256
This is dumb:

```
>>> from datetime import datetime
>>> np.datetime64(datetime.now()).astype((str, 100))
'2017-09-18T23:12:59.814390'
>>> np.datetime64(datetime.now()).astype((str, 256))
''
```

### PR DESCRIPTION
**Pull request by eric-wieser**
Fixes #9712

### CODE REVIEW

File: numpy/core/tests/test_datetime.py [hunk 1]
```diff
@@ -518,31 +518,38 @@ def test_pydatetime_creation(self):
...
+
         # Datetime to string
         assert_equal(str_a, dt_a.astype('S0'))
+        str_b = np.empty_like(str_a)
         str_b[...] = dt_a
         assert_equal(str_a, str_b)
 
-        # Convert the 'S' to 'U'
-        str_a = str_a.astype('U')
-        str_b = str_b.astype('U')
```
Comment (eric-wieser) [inline numpy/core/tests/test_datetime.py]: This wasn't a good test, as it would still pass if the `str_b[...] = ...` below was a no op!

**Comment by eric-wieser** [comment]
Darn warnings that aren't noticed by `runtests.py`...

**Review by mhvk — APPROVED** [review]
Looks good to me. And nice catch on the use of `char` for a length.

**Comment by charris** [comment]
Thanks Eric.

**Pull request merged**